universal-application-tool-0.0.1/app/assets/javascripts/admin_validation.ts

```diff
@@ -0,0 +1,74 @@
+/** The validation controller provides basic client-side validation of admin form fields. */
+class AdminValidationController {
+  static readonly MULTI_OPTION_QUESTION_FIELD_NAME_CREATE = '#question-settings input[name="newOptions[]"]';
+  static readonly MULTI_OPTION_QUESTION_FIELD_NAME_EDIT = '#question-settings input[name="options[]"]';
+
+  isMultiOptionCreateValid = true;
+  isMultiOptionEditValid = true;
+
+  constructor() {
+    // Attach listener to admin program edit form.
+    const adminProgramEditForm = document.getElementById("question-form");
+    if (adminProgramEditForm) {
+      adminProgramEditForm.addEventListener("submit", (event) => { return this.attemptSubmit(event); });
+    }
+  }
+
+  private attemptSubmit(event: Event): boolean {
+    this.checkFields();
+    if (!this.isValid()) {
+      event.preventDefault();
+      return false;
+    }
+    return true;
+  }
+
+  private checkFields() {
+    this.isMultiOptionCreateValid = this.validateMultiOptionQuestionCreate();
+    this.isMultiOptionEditValid = this.validateMultiOptionQuestionEdit();
+  }
+
+  updateFieldErrorState(element: Element, fieldName: string, isValid: boolean) {
+    const errorDiv = element.parentElement ? element.parentElement.querySelector(fieldName + '-error') : element.querySelector(fieldName + '-error');

```

Comment:
question (blocking): when does the element not have a parent?

Reply:
Good point -- I think we were trying to catch a case that actually doesn't exist. I deleted this -- but...

@dkatzz please take a look at this because you wrote this line of code and I don't want to delete logic that you think is important.

Reply:
Yeah, the compiler complained about it possibly being null when doing `element.parentElement.querySelector...`, but the exclamation point is a good way to make this clearer - thanks!

Reply:
That works if you're super confident that `element` will always be the element you're expecting. Otherwise could add some assertions that throw descriptive errors if it's not.